pkg/runtime/scheduler/streamer.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2024 The Dapr Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package scheduler
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+
+	"github.com/golang/protobuf/ptypes/wrappers"
+	"google.golang.org/grpc/codes"
+
+	"github.com/dapr/dapr/pkg/actors"
+	invokev1 "github.com/dapr/dapr/pkg/messaging/v1"
+	internalv1pb "github.com/dapr/dapr/pkg/proto/internals/v1"
+	schedulerv1pb "github.com/dapr/dapr/pkg/proto/scheduler/v1"
+	"github.com/dapr/dapr/pkg/runtime/channels"
+	"github.com/dapr/kit/concurrency"
+)
+
+type streamer struct {
+	stream   schedulerv1pb.Scheduler_WatchJobsClient
+	resultCh chan *schedulerv1pb.WatchJobsRequest
+
+	actors   actors.ActorRuntime
+	channels *channels.Channels
+
+	wg sync.WaitGroup
+}
+
+func (s *streamer) run(ctx context.Context) error {
+	return concurrency.NewRunnerManager(s.receive, s.outgoing).Run(ctx)
+}
+
+func (s *streamer) receive(ctx context.Context) error {
+	defer s.wg.Wait()
+
+	for {
+		resp, err := s.stream.Recv()
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if err != nil {
```

Comment:
Should we add a check in here if the `err == io.EOF` and return nil? Meaning the stream ended, no err, or you will want to return err here in that case?

I had it working that if the scheduler went down that it would keep retrying to connect in the background bc if there are addresses it should connect in theory. Assuming Im following these changes correctly, I believe the logic in this PR now bubbles up the err to `runtime.go` and can cause daprd to go down if watchJobs returns an err. Did you test for that case locally to ensure daprd keeps retrying?

Reply:
Thanks for the catch @cicoyle! I've not updated with a `reconnect` integration test to ensure re-connection works :slightly_smiling_face: